File: Algorithm/TwoPointers/atcoder/B13/B13.go
// 以下に、**Go (v1.20.6)** を用いた、しゃくとり法（二重ポインタ）による高速解法を提示します。
// 時間・メモリ効率を意識しており、**`O(N)` 時間、`O(N)` メモリ**で動作します。

// ---

// ## ✅ 問題再掲（要約）

// * 配列 `A[0...N-1]` の連続部分列のうち、**和が K 以下**の個数を数える。
// * 制約：`1 ≤ N ≤ 10^5`, `1 ≤ K, A[i] ≤ 10^9`
// * 実行時間制限：1 秒、メモリ制限：1024 MiB

// ---

// ## ✅ Go 実装（型明示・高速処理）

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	// 高速読み取りのため bufio.Scanner 使用
	reader := bufio.NewReader(os.Stdin)

	// 1行目：N, K
	line1, _ := reader.ReadString('\n')
	parts := strings.Fields(line1)
	N, _ := strconv.Atoi(parts[0])
	K, _ := strconv.Atoi(parts[1])

	// 2行目：A[0..N-1]
	line2, _ := reader.ReadString('\n')
	aParts := strings.Fields(line2)
	A := make([]int, N)
	for i := 0; i < N; i++ {
		A[i], _ = strconv.Atoi(aParts[i])
	}

	// Two pointers（しゃくとり法）
	var count int = 0
	var sum int = 0
	var right int = 0

	for left := 0; left < N; left++ {
		for right < N && sum+A[right] <= K {
			sum += A[right]
			right++
		}

		count += right - left

		if right == left {
			right++
		} else {
			sum -= A[left]
		}
	}

	fmt.Println(count)
}

// ## ✅ 処理内容と型の明示

// | 変数名                             | 型       | 説明           |
// | ------------------------------- | ------- | ------------ |
// | `N`, `K`                        | `int`   | 入力サイズ・上限     |
// | `A`                             | `[]int` | 配列（最大10⁵要素）  |
// | `sum`, `count`, `left`, `right` | `int`   | 計算用ポインタ・累積変数 |

// ---

// ## ✅ パフォーマンスとリソース

// | 項目    | 評価                           |
// | ----- | ---------------------------- |
// | 時間計算量 | `O(N)`                       |
// | 空間計算量 | `O(N)`（配列Aのみ）                |
// | メモリ目安 | ≒ 400KB（`10⁵ × 4B`）          |
// | 実行速度  | 数十ms（`go run` or `go build`） |

// ---

// ## ✅ 入力例

// ```text
// 7 50
// 11 12 16 22 27 28 31
// ```

// ```bash
// go run main.go < input.txt
// ```

// 出力：

// ```
// 13
// ```

// ---

// ## ✅ 処理フロー概要（図解）

// ```
// [11, 12, 16, 22, 27, 28, 31]
//  ↑                            left
//  ↑                            right
// sum = 0 → 右へ進めて、合計 ≤ K の最大範囲 → count += (right - left)
// → left を進めて同様に処理
// ```

// ---

// ## ✅ 補足：高速 I/O の工夫

// * `bufio.NewReader` で1行ずつ高速読み込み
// * `strings.Fields()` により空白文字を自動分割（タブ・空白など）
// * すべての `Atoi` は `int` 型で扱い、64bit対応不要（Go の `int` は 64bit）


// 複数行に渡る A\[i] 入力形式や、ベンチマーク用コードの追加、`bufio.Scanner` 使用への切り替えも可能です。

File: Algorithm/TwoPointers/atcoder/B13/B13.ts
// * 入力は `fs.readFileSync` を使用して一括で読み取り、
// * アルゴリズムは **しゃくとり法（Two Pointers）** を用いて、計算量 `O(N)` で解を導きます。
// * 実行時間・メモリともに制限内（1秒 / 1024MiB）に収まるように実装しています。

// ---

// ### ✅ TypeScript 実装（`fs`使用・最適化済）

import * as fs from 'fs';

// 入力取得・分割
const input: string[] = fs.readFileSync('/dev/stdin', 'utf8').trim().split('\n');
const [N, K]: number[] = input[0].split(' ').map(Number);
const A: number[] = input[1].split(' ').map(Number);

// Two Pointers（しゃくとり法）で全連続部分列のうち合計 <= K の個数を数える
let count: number = 0;
let sum: number = 0;
let right: number = 0;

for (let left = 0; left < N; left++) {
    while (right < N && sum + A[right] <= K) {
        sum += A[right];
        right++;
    }

    count += right - left;

    if (right === left) {
        // A[left] > K のとき、right も一緒に進める
        right++;
    } else {
        sum -= A[left];
    }
}

console.log(count);

// ### ✅ 入力例

// ```
// 7 50
// 11 12 16 22 27 28 31
// ```

// 出力：

// ```
// 13
// ```

// ---

// ### ✅ 工夫ポイント

// | 項目     | 内容                                                   |
// | ------ | ---------------------------------------------------- |
// | 入力     | `fs.readFileSync("/dev/stdin", "utf8")` により高速な一括読み取り |
// | アルゴリズム | `O(N)` のしゃくとり法で、重複計算回避                               |
// | メモリ    | `A[]`と定数変数のみで、メモリ使用量は`O(N)`                          |
// | 型      | `number[]`や`number`を明示し、TypeScriptでの型安全確保            |

// ---

// ### ⏱ パフォーマンス評価（目安）

// | 項目     | 規模     | 備考              |
// | ------ | ------ | --------------- |
// | 最大 N   | 10⁵    | ループ最大 2×10⁵ 回   |
// | 実行時間   | ≒ 数十ms | 1秒以内で十分に収まる     |
// | メモリ使用量 | ≒ 数MB  | 1024MiB 制限に余裕あり |
